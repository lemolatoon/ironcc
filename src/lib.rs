pub mod analyze;
pub mod generate;
pub mod parse;
pub mod tokenize;
