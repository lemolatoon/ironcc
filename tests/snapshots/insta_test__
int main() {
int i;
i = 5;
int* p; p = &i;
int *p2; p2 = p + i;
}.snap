---
source: src/lib.rs
expression: parse_program(src)
---
Program {
    components: [
        Func(
            Declaration {
                ty_spec: Int,
                n_star: 0,
                declrtr: Func(
                    Ident(
                        "main",
                    ),
                    [],
                ),
                pos: Position {
                    n_char: 0,
                    n_line: 1,
                },
            },
            Stmt {
                kind: Block(
                    [
                        Stmt {
                            kind: Declare(
                                Declaration {
                                    ty_spec: Int,
                                    n_star: 0,
                                    declrtr: Ident(
                                        "i",
                                    ),
                                    pos: Position {
                                        n_char: 0,
                                        n_line: 2,
                                    },
                                },
                            ),
                        },
                        Stmt {
                            kind: Expr(
                                Expr {
                                    kind: Assign(
                                        Expr {
                                            kind: LVar(
                                                "i",
                                            ),
                                            pos: Position {
                                                n_char: 0,
                                                n_line: 3,
                                            },
                                        },
                                        Expr {
                                            kind: Num(
                                                5,
                                            ),
                                            pos: Position {
                                                n_char: 4,
                                                n_line: 3,
                                            },
                                        },
                                    ),
                                    pos: Position {
                                        n_char: 2,
                                        n_line: 3,
                                    },
                                },
                            ),
                        },
                        Stmt {
                            kind: Declare(
                                Declaration {
                                    ty_spec: Int,
                                    n_star: 1,
                                    declrtr: Ident(
                                        "p",
                                    ),
                                    pos: Position {
                                        n_char: 0,
                                        n_line: 4,
                                    },
                                },
                            ),
                        },
                        Stmt {
                            kind: Expr(
                                Expr {
                                    kind: Assign(
                                        Expr {
                                            kind: LVar(
                                                "p",
                                            ),
                                            pos: Position {
                                                n_char: 8,
                                                n_line: 4,
                                            },
                                        },
                                        Expr {
                                            kind: Addr(
                                                Expr {
                                                    kind: LVar(
                                                        "i",
                                                    ),
                                                    pos: Position {
                                                        n_char: 13,
                                                        n_line: 4,
                                                    },
                                                },
                                            ),
                                            pos: Position {
                                                n_char: 12,
                                                n_line: 4,
                                            },
                                        },
                                    ),
                                    pos: Position {
                                        n_char: 10,
                                        n_line: 4,
                                    },
                                },
                            ),
                        },
                        Stmt {
                            kind: Declare(
                                Declaration {
                                    ty_spec: Int,
                                    n_star: 1,
                                    declrtr: Ident(
                                        "p2",
                                    ),
                                    pos: Position {
                                        n_char: 0,
                                        n_line: 5,
                                    },
                                },
                            ),
                        },
                        Stmt {
                            kind: Expr(
                                Expr {
                                    kind: Assign(
                                        Expr {
                                            kind: LVar(
                                                "p2",
                                            ),
                                            pos: Position {
                                                n_char: 9,
                                                n_line: 5,
                                            },
                                        },
                                        Expr {
                                            kind: Binary(
                                                Binary {
                                                    kind: Add,
                                                    lhs: Expr {
                                                        kind: LVar(
                                                            "p",
                                                        ),
                                                        pos: Position {
                                                            n_char: 14,
                                                            n_line: 5,
                                                        },
                                                    },
                                                    rhs: Expr {
                                                        kind: LVar(
                                                            "i",
                                                        ),
                                                        pos: Position {
                                                            n_char: 18,
                                                            n_line: 5,
                                                        },
                                                    },
                                                },
                                            ),
                                            pos: Position {
                                                n_char: 16,
                                                n_line: 5,
                                            },
                                        },
                                    ),
                                    pos: Position {
                                        n_char: 12,
                                        n_line: 5,
                                    },
                                },
                            ),
                        },
                    ],
                ),
            },
        ),
    ],
}
